cipd_packages/device_doctor/lib/src/android_device.dart

```diff
@@ -380,7 +381,54 @@ class AndroidDevice implements Device {
 
   @override
   Future<void> recover() async {
-    await eval('adb', <String>['-s', deviceId!, 'reboot'], canFail: false);
+    await cleanDevice();
+  }
+
+  @visibleForTesting
+  Future<void> waitForDeviceUp({ProcessManager? processManager}) async {
+    processManager ??= LocalProcessManager();
+    final List<String> waitForDeviceArgs = [
+      'shell',
+      'true',
+    ];
+    final int maxTries = 4;
+    final int sleepSecs = 5;
+    for (var i = 0; i < maxTries; i++) {
+      final Process proc = await startProcess('adb', waitForDeviceArgs, processManager: processManager);
+      if (0 == await proc.exitCode) {
+        break;
+      }
+      await Future.delayed(Duration(seconds: sleepSecs));
```

Comment:
I think you'll need to use `Future.wait`, otherwise your `Future.delayed` will never time out your `adb` process if it's hanging:

```suggestion
    bool success = false;
    for (var i = 0; i < maxTries; i++) {
      final Process proc = await startProcess('adb', waitForDeviceArgs, processManager: processManager);
      if (0 == await proc.exitCode) {
        break;
      }
      await Future.delayed(Duration(seconds: sleepSecs));
      await Future.wait(<Future<void>>[
        proc.exitCode.then((int code) {
          if (code == 0) success = true;
        }),
        Future.delayed(Duration(seconds: sleepSecs)),
      ]);
      if (sucess) {
        break;
      }
```

Reply:
oh wait, I was wrong, you ACTUALLY want Future.any: https://api.dart.dev/stable/3.1.2/dart-async/Future/any.html

Future.wait will wait for ALL of your futures, Future.any any one of them.